tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+
+        try:
+            # enable the exthandlers_handler to interact with every actor before it is exposed.
+            for actor in protocol._actors:
+                patched_popen = actor.patch_popen() # We delay creation so that *this* patched_popen can "see" the prior ones (they must be started already).
+                patched_popen.start() # Allow later patches to "see" this one (as subprocess.Popen).
+
+                protocol._patched_popens.append(patched_popen)
+
+            yield protocol, get_exthandlers_handler(protocol)
+
+        finally:
+            for patched_popen in protocol._patched_popens:
+                patched_popen.stop()
+
+
+def add_extension_actors(protocol, incarnation, *actors):
+    """
+    Given a protocol obj returned by a extension_actor.get_protocol_and_handler(firstActor, *remainingActors) call, injects the
+    extensions emulated by (the objects within) actors.
+    
+    It accomplishes this though applying the following updates onto the provided protocol:
+        *   adds unique data sources for *.zip extension resources garnered from actors (unique meaning ones not already given
+            by [firstActor, *remainingActors], if any)
+        *   updates ext_conf to reflect the extension names and versions within actors
+        *   updates the goal state's incarnation (to force the agent to re-read it)
+
+    In order to force a goal state update, the incarnation parameter needs to be greater than the value passed in the last
+    extension_actor.update_extension_actors() function call, or greater than 0 if this is the first of such calls.
+
+    Note that non-default manifest.xml files must use the same url format as the default in order to be compatible with this
+    function. Specifically, the *.zip resource urls for an extension must end in "{ExtensionName}__{ExtensionVersion}".
+
+    Note that this function does not update xml for ext_conf or manifest, meaning that added extensions must already be present
+    within those files as provided by the firstActor's data source. This could probably be implemented with the
+    WireProtocolDataBase.replace_xml_element_value function, but it would probably make sense to first implement autogeneration
+    of those files within ExtensionActor.__init__, as we would then know exactly the xml elements in those files. This would
+    simplify the logic of adding the new actors, but is a bigger change. Ultimately, this feature isn't currently (i.e. as of
+    writing) needed.
+    """
+
+    new_actors = list(filter(lambda actor: actor not in protocol._actors, actors))  # list() applies the lambda before we edit protocol._actors
+    protocol._actors.extend(new_actors)
+
+    for actor in new_actors:
+        patched_popen = actor.patch_popen() # Like in get_protocol_and_handler, the patch_popen() and start() calls must
+        patched_popen.start()               # be interleaved.
```

Comment:
where do we stop the mock?

Reply:
Let me know if you just missed this comment, or if you saw it but it wasn't clear enough:

line 248: `# Keep track of the patches so the finally clause within get_protocol_and_handler can stop them.`
line 249: `protocol._patched_popens.append(patched_popen)`

and the finally clause referenced in the comment (line 211-214):
```
finally:
    for patched_popen in protocol._patched_popens:
        patched_popen.stop()
```

To answer your question, we're keeping track of all of the mock popen objects, and stopping them when the `with get_protocol_and_handler` goes out of scope.

Reply:
I see; the cleanup is on a different function. Can we improve this (maybe use a context manager) so  that initialization/cleanup are together?

Reply:
Just so we're on the same page, you mean make the `add_extension_actors` function a context manager?